p4_symbolic/symbolic/parser.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Hardcodes the behavior of an interesting p4 parser that is part
+// of the p4 program we are interested in.
+// The hardcoded behavior sets the validity of certain header fields
+// based on the fields in the packet, and sets the default values
+// for local_metadata fields.
+
+#include "p4_symbolic/symbolic/parser.h"
+
+#include "z3++.h"
+
+namespace p4_symbolic {
+namespace symbolic {
+namespace parser {
+
+gutil::StatusOr<std::vector<z3::expr>> EvaluateHardcodedParser(
```

Comment:
Instead of having this function both modify the context, and return constraints, can we make it uniform, and just return constraints. So, for example, call the function `HardcodedParserConstraints(const SymbolicPerPacketState& state)`

Reply:
That works, it will make the code slightly more complicated. The other way definitely wont work (we need to return some non-assignable constraints). I will do this.